File: examples/typescript/src/order-total.ts
export interface LineItem {
  unitPriceCents: number;
  quantity: number;
}

export function calculateOrderTotal(items: readonly LineItem[]): number {
  return items.reduce((total, item) => {
    if (!Number.isSafeInteger(item.unitPriceCents) || item.unitPriceCents < 0) {
      throw new RangeError("unitPriceCents must be a non-negative integer");
    }

    if (!Number.isSafeInteger(item.quantity) || item.quantity <= 0) {
      throw new RangeError("quantity must be a positive integer");
    }

    const lineTotal = item.unitPriceCents * item.quantity;
    const nextTotal = total + lineTotal;

    if (!Number.isSafeInteger(nextTotal)) {
      throw new RangeError("order total exceeds the safe integer range");
    }

    return nextTotal;
  }, 0);
}

File: examples/go/order_total.go
package ordertotal

import (
	"errors"
	"math"
)

var (
	ErrInvalidUnitPrice = errors.New("unit price must be non-negative")
	ErrInvalidQuantity  = errors.New("quantity must be positive")
	ErrTotalOverflow    = errors.New("order total exceeds the int64 range")
)

type LineItem struct {
	UnitPriceCents int64
	Quantity       int64
}

func CalculateOrderTotal(items []LineItem) (int64, error) {
	var total int64

	for _, item := range items {
		if item.UnitPriceCents < 0 {
			return 0, ErrInvalidUnitPrice
		}
		if item.Quantity <= 0 {
			return 0, ErrInvalidQuantity
		}
		if item.UnitPriceCents > math.MaxInt64/item.Quantity {
			return 0, ErrTotalOverflow
		}

		lineTotal := item.UnitPriceCents * item.Quantity
		if total > math.MaxInt64-lineTotal {
			return 0, ErrTotalOverflow
		}
		total += lineTotal
	}

	return total, nil
}
